src/ray/object_manager/object_manager.cc

```diff
@@ -349,17 +350,132 @@ ray::Status ObjectManager::SendObjectData(const ObjectID &object_id,
+  }
+
+  // Initialize fields.
+  active_wait_requests_.emplace(wait_id, WaitState(*main_service_, wait_ms, callback));
+  auto &wait_state = active_wait_requests_.find(wait_id)->second;
+  wait_state.object_id_order = object_ids;
+  wait_state.wait_ms = wait_ms;
+  wait_state.num_required_objects = num_required_objects;
+  for (auto &oid : object_ids) {
+    if (local_objects_.count(oid) > 0) {
+      wait_state.found.insert(oid);
+    } else {
+      wait_state.remaining.insert(oid);
+    }
+  }
+
+  if (wait_state.remaining.empty()) {
+    WaitComplete(wait_id);
+  } else {
+    for (auto &oid : wait_state.remaining) {
+      // Lookup remaining objects.
+      wait_state.requested_objects.insert(oid);
+      RAY_CHECK_OK(object_directory_->LookupLocations(
+          oid, [this, wait_id](const std::vector<ClientID> &client_ids,
+                               const ObjectID &object_id) {
+            auto &wait_state = active_wait_requests_.find(wait_id)->second;
+            if (!client_ids.empty()) {
+              wait_state.remaining.erase(object_id);
+              wait_state.found.insert(object_id);
+            }
+            wait_state.requested_objects.erase(object_id);
+            if (wait_state.requested_objects.empty()) {
+              WaitLookupComplete(wait_id);
+            }
+          }));
+    }
+  }
   return ray::Status::OK();
 }
 
+void ObjectManager::WaitLookupComplete(const UniqueID &wait_id) {
+  auto &wait_state = active_wait_requests_.find(wait_id)->second;
+  if (wait_state.found.size() >= wait_state.num_required_objects ||
+      wait_state.wait_ms == 0) {
+    // Requirements already satisfied.
+    WaitComplete(wait_id);
+  } else {
+    for (auto &oid : wait_state.remaining) {
+      // Subscribe to object notifications.
+      wait_state.requested_objects.insert(oid);
+      RAY_CHECK_OK(object_directory_->SubscribeObjectLocations(
+          wait_id, oid, [this, wait_id](const std::vector<ClientID> &client_ids,
+                                        const ObjectID &object_id) {
+            auto &wait_state = active_wait_requests_.find(wait_id)->second;
+            if (wait_state.remaining.count(object_id) != 0) {
+              wait_state.remaining.erase(object_id);
+              wait_state.found.insert(object_id);
+            }
+            wait_state.requested_objects.erase(object_id);
+            RAY_CHECK_OK(
+                object_directory_->UnsubscribeObjectLocations(wait_id, object_id));
+            if (wait_state.found.size() >= wait_state.num_required_objects) {
+              WaitComplete(wait_id);
+            }
+          }));
+    }
+    // Set timeout.
+    // TODO (hme): If we need to just wait for all objects independent of time
+    // (i.e. infinite wait time), determine what the value of wait_ms should be and
+    // skip this call. WaitComplete will be invoked when all objects have locations.
+    wait_state.timeout_timer->async_wait(
```

Comment:
There may be some issues with the structure of the code.

For example, `Wait` will lead to one call to `WaitLookupComplete` for each requested object, and `WaitLookupComplete` will start the `timeout_timer`. But it only makes sense to start that timer once, (probably in `Wait`).

Reply:
`WaitLookupComplete` is invoked once for each invocation to `Wait`; I'll update the name of this method to `AllWaitLookupsComplete` make it easier to read.